prov/rxm/src/rxm_cq.c

```diff
@@ -759,6 +761,211 @@ static int rxm_handle_remote_write(struct rxm_ep *rxm_ep,
 	return 0;
 }
 
+static inline void rxm_ep_format_atomic_resp_pkt_hdr(struct rxm_conn *rxm_conn,
+				struct rxm_tx_atomic_buf *tx_buf,
+				size_t data_len, uint32_t pkt_op,
+				enum fi_datatype datatype, uint8_t atomic_op)
+{
+	rxm_ep_format_tx_buf_pkt(rxm_conn, data_len, pkt_op, 0, 0, 0,
+				 &tx_buf->pkt);
+	tx_buf->pkt.ctrl_hdr.type = ofi_ctrl_atomic_resp;
+	tx_buf->pkt.hdr.op = pkt_op;
+	tx_buf->pkt.hdr.atomic.datatype = datatype;
+	tx_buf->pkt.hdr.atomic.op = atomic_op;
+	tx_buf->pkt.hdr.atomic.ioc_count = 0;
+}
+
+static ssize_t rxm_atomic_send_resp(struct rxm_ep *rxm_ep,
+				    struct rxm_rx_buf *rx_buf,
+				    struct rxm_tx_atomic_buf *resp_buf,
+				    ssize_t result_len, uint32_t status)
+{
+	struct rxm_deferred_tx_entry *def_tx_entry;
+	struct rxm_atomic_resp_hdr *atomic_hdr;
+	ssize_t ret;
+	ssize_t resp_len = result_len + sizeof(struct rxm_atomic_resp_hdr) +
+				sizeof(struct rxm_pkt);
+
+	resp_buf->hdr.state = RXM_ATOMIC_RESP_SENT;
+	rxm_ep_format_atomic_resp_pkt_hdr(rx_buf->conn,
+					  resp_buf,
+					  resp_len,
+					  rx_buf->pkt.hdr.op,
+					  rx_buf->pkt.hdr.atomic.datatype,
+					  rx_buf->pkt.hdr.atomic.op);
+	resp_buf->pkt.ctrl_hdr.conn_id = rx_buf->conn->handle.remote_key;
+	resp_buf->pkt.ctrl_hdr.msg_id = rx_buf->pkt.ctrl_hdr.msg_id;
+	atomic_hdr = (struct rxm_atomic_resp_hdr *) resp_buf->pkt.data;
+	atomic_hdr->status = htonl(status);
+	atomic_hdr->result_len = htonl(result_len);
+
+	ret = fi_send(rx_buf->conn->msg_ep, &resp_buf->pkt, resp_len,
+		      resp_buf->hdr.desc, 0, resp_buf);
+	if (OFI_UNLIKELY(ret)) {
+		FI_WARN(&rxm_prov, FI_LOG_CQ,
+			"Unable to send Atomic Response\n");
+		if (OFI_LIKELY(ret == -FI_EAGAIN)) {
```

Comment:
Should we be checking the resource management flags of the domain? The user might not request RM, which might mean that we can kick the request entirely rather than internally queuing it.

Reply:
That is an interesting point, I'm not sure what the correct answer is. What is happing here is that if the internal emulated atomic protocol can not send a protocol Atomic Response Ack/Nak immediately, we queue it to be sent when it can. This call is transparent to the application, and the sender will hang if Atomic Response is not ultimately received.

Reply:
Not terribly important, but Sean might bring it up since the provider seems to be assuming custody of the transaction as soon as we put it on the deferred queue.